CovertInfiltration/Src/CovertInfiltration/Classes/XComGameState_SquadPickupPoint.uc

```diff
@@ -0,0 +1,218 @@
+	local XComGameState NewGameState;
+
+
+	NewGameState = class'XComGameStateContext_ChangeContainer'.static.CreateChangeState("CI: Exfiltration Confirmed");
+	PickupPoint = XComGameState_SquadPickupPoint(NewGameState.ModifyStateObject(class'XComGameState_SquadPickupPoint', GetReference().ObjectID));
+	XComHQ = XComGameState_HeadquartersXCom(NewGameState.ModifyStateObject(class'XComGameState_HeadquartersXCom', XComHQ.GetReference().ObjectID));
+
+	CostScalars.Length = 0;
+	PickupPoint.bSquadExfiltrated = true;
+
+	`XCOMHQ.PayStrategyCost(NewGameState, ExfiltrateCost, CostScalars);
+	ClearUnitsFromAction(NewGameState);
+	
+	`XCOMGAME.GameRuleset.SubmitGameState(NewGameState);
+	
+	DestroyTheEvidence();
+
+	`XSTRATEGYSOUNDMGR.PlayGeoscapeMusic();
+	InteractionComplete(true);
+}
+
+simulated function ClearUnitsFromAction(XComGameState NewGameState)
+{
+	local XComGameState_CovertAction CovertAction;
+	local XComGameState_StaffSlot SlotState;
+	local XComGameState_Unit Unit;
+	local XComGameStateHistory History;
+	local CovertActionStaffSlot CovertActionSlot;
+
+	History = `XCOMHISTORY;
+	CovertAction = XComGameState_CovertAction(History.GetGameStateForObjectID(ActionRef.ObjectID));
+
+	foreach CovertAction.StaffSlots(CovertActionSlot)
+	{
+		SlotState = XComGameState_StaffSlot(History.GetGameStateForObjectID(CovertActionSlot.StaffSlotRef.ObjectID));
+		if (SlotState.IsSlotFilled())
+		{
+			Unit = SlotState.GetAssignedStaff();
+			if (Unit.IsSoldier() && Unit.UsesWillSystem())
+			{
+				class'X2Helper_Infiltration'.static.CreateWillRecoveryProject(NewGameState, Unit);
+			}
+			SlotState.EmptySlot(NewGameState);
+		}
+	}
+}
+
+function DestroyTheEvidence()
+{
+	local XComGameState_SquadPickupPoint PickupPoint;
+	local XComGameState_CovertAction CovertAction;
+	local XComGameState NewGameState;
+
+	NewGameState = class'XComGameStateContext_ChangeContainer'.static.CreateChangeState("CI: SkyrangerExfiltrate Cleanup");
+
+	PickupPoint = XComGameState_SquadPickupPoint(NewGameState.ModifyStateObject(class'XComGameState_SquadPickupPoint', GetReference().ObjectID));
+	PickupPoint.bConsumed = true;
+	PickupPoint.RemoveEntity(NewGameState);
+
+	if (bSquadExfiltrated)
+	{
+		CovertAction = XComGameState_CovertAction(NewGameState.ModifyStateObject(class'XComGameState_CovertAction', ActionRef.ObjectID));
+		CovertAction.RemoveEntity(NewGameState);
+	}
+
+	`XCOMGAME.GameRuleset.SubmitGameState(NewGameState);
+}
+
+static function Purge()
+{
+	local array<XComGameState_SquadPickupPoint> PickupPoints;
+	local XComGameState_SquadPickupPoint PickupPoint;
+	local XComGameStateHistory History;
+	local XComGameState NewGameState;
+	local bool bDirty;
+
+	History = `XCOMHISTORY;
+	NewGameState = class'XComGameStateContext_ChangeContainer'.static.CreateChangeState("CI: Purging SquadPickupPoint(s)");
+
+	foreach History.IterateByClassType(class'XComGameState_SquadPickupPoint', PickupPoint)
+	{
+		PickupPoints.AddItem(PickupPoint);
```

Comment:
Not sure why you marked conversation as resolved - you are still using 2 `foreach`s and an intermediary array - just do lines 190 and 191 (+ `bDirty = true`) inside the history foreach